src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,48 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> top;
+    private int size;
+
+    private static class Node<T> {
+        private final T value;
+        private final Node<T> previous;
 
+        public Node(T value, Node<T> previous) {
+            this.value = value;
+            this.previous = previous;
+        }
+    }
+
+    public void push(T value) {
+        if (size == 0) {
+            top = new Node<>(value, null);
+        } else {
+            top = new Node<>(value, top);
+        }
+        size++;
     }
 
     public T peek() {
-        return null;
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+        return top.value;
     }
 
     public T pop() {
-        return null;
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+        Node<T> requiredNode = top;
```

Comment:
you can simply call the Peek() method here.
```suggestion
        Node<T> requiredNode = Peek();
```

Reply:
done